feat(Response): add fromDiskCache/fromServiceWorker info

ebidel opened the pull request:
Fixes #1551

I tried to use `one-style.html` for the test but it would require additions to sw.js and the test server doesn't support caching AFAIK.

ebidel committed:
feat(Response): add fromDiskCache/fromServiceWorker info

ebidel committed:
doc linting

ebidel commented:
I'm not able to run all the tests b/c of https://github.com/GoogleChrome/puppeteer/issues/1600. Any idea why Travis is unhappy?

aslushnikov commented on docs/api.md:
```diff
@@ -182,6 +182,8 @@
   * [request.url](#requesturl)
 - [class: Response](#class-response)
   * [response.buffer()](#responsebuffer)
+  * [response.fromDiskCache](#responsefromdiskcache)
```
If I understand it correctly, every response could be served either from network, disk cache or service worker. So it's never the case that both `fromDiskCache` and `fromServiceWorker` are `true`. Is this the case? 

If so, I'd rather do a `response.servedFrom()` method that returns either `network`, `diskcache` or `serviceworker`.

amprew replied:
They can both be true at the same time. If you check a request against `https://googlechrome.github.io/samples/service-worker/basic/`, it shows both `fromDiskCache` and `fromServiceWorker` as true. From what I understand `fromServiceWorker` is when the request is served from the `fetch` event inside the web worker. But fromDiskCache will also return true if you return from the Cache API.

Hope I haven't misunderstood your question.

ebidel replied:
It looks like they're different. Running against chromestatus shows all 4 permutations can happen (urls have been truncated):

<img width="733" alt="screen shot 2017-12-16 at 8 43 22 am" src="https://user-images.githubusercontent.com/238208/34072496-466c36f4-e23d-11e7-829f-c0c5f8b7eec3.png">

That makes sense. The disk cache is different than the sw cache storage api.

amprew commented on test/test.js:
```diff
@@ -3301,6 +3303,24 @@ describe('Page', function() {
+      expect(responses[0].fromDiskCache).toBe(false);
+      expect(responses[0].fromServiceWorker).toBe(false);
+      responses = [];
+      response = await page.reload({waitUntil: 'networkidle2'});
```
looks like `response` is being assigned to the scope but not used

ebidel replied:
done. weird linting didnt catch this.

ebidel committed:
Feedback

ebidel committed:
Fix test

aslushnikov commented on test/test.js:
```diff
@@ -3301,6 +3303,30 @@ describe('Page', function() {
   });
+
+  describe('Response', function() {
+    it('specifies when response comes from cache and service worker', async() => {
```
page and browser are passed into the test:

```js
async ({page, browser}) => {
  // ..
});
```

No need to create another one; reusing browser speeds up tests significantly

ebidel replied:
I was doing that but it was causing other tests to fail: https://travis-ci.org/GoogleChrome/puppeteer/jobs/317396566#L556

aslushnikov commented on test/test.js:
```diff
@@ -3301,6 +3303,30 @@ describe('Page', function() {
+          responses.push(response);
+      });
+
+      await newPage.goto('https://www.chromestatus.com/features', {waitUntil: 'networkidle2'});
```
please don't use external URLs; we aim for the hermetic builds.

ebidel replied:
> I tried to use one-style.html for the test but it would require additions to sw.js and the test server doesn't support caching AFAIK.

What do you suggest for the server part?

ebidel replied:
I've added caching to the test server. Still seeing a lot of unrelated errors that I could use your help on. Not familiar enough with the worker and custom test server.

aslushnikov commented on docs/api.md:
```diff
@@ -2065,6 +2067,16 @@ Contains the URL of the request.
 #### response.buffer()
 - returns: <Promise<[Buffer]>> Promise which resolves to a buffer with response body.
 
+#### response.fromDiskCache
```
let's make them functions to unify our API

ebidel replied:
done

ebidel committed:
Use methods instead of props
- add browser caching to test server

ebidel committed:
Fixes after rebase

ebidel committed:
newlines

ebidel commented:
Rebased. Still need guidance on what do about https://github.com/GoogleChrome/puppeteer/pull/1601#discussion_r157564662. The added caching to the server causes unrelated tests to fail. 

Instead, we could go back to a well-known URL that has caching + sw. https://googlechrome.github.io/samples/service-worker/basic/index.html is our official service worker demo and won't be changing.

ebidel commented:
I can also switch to `<link rel="stylesheet" href="https://developers.google.com/web/styles/wf-root.css">` in sw-fetch.html if you don't mind a page that uses an external resource. That makes all tests pass :)

ebidel commented:
rebased

aslushnikov commented:
@ebidel let me introduce caching to the simple server so that this can be merged in.

aslushnikov commented:
@ebidel the patch doesn't handle memory caching, and it's quite hard to predict if the response will be memory-cached or disk-cached. As far as I know the behavior is not spec'ed in chrome.

I'd expose `response.fromCache()` instead of `response.fromDiskCache()` that will report `true` if the response is served either from memory or disk cache. wdyt? I have a draft that implements this.

ebidel commented:
I don't mind changing the name but doesn't the protocol differentiate the two? The name and docs from [`Response.fromDiskCache`](https://chromedevtools.github.io/devtools-protocol/tot/Network/#type-Response) suggests that it's just disk. Or is it a deceiving name?

aslushnikov commented:
@ebidel the protocol differentiates between the two; memory-cached responses are reported with `Network.requestServedFromCache`. It's hard to tell when chrome will use memory cache and when disk cache, as far as i know the behavior is not spec'ed.

ebidel commented:
That must be how the frontend shows each:

<img width="1169" alt="screen shot 2018-01-05 at 11 01 15 am" src="https://user-images.githubusercontent.com/238208/34623882-e165b264-f207-11e7-8aeb-bd175236562e.png">

Does DT prioritize showing "from memory cache" if the resource is in both? Should we do the same in PPTR or add `response.fromMemoryCache()` somehow?

ebidel commented:
What about this:

```js
const responsesFromMemCache = [];
page._client.on('Network.requestServedFromCache', result => {
  responsesFromMemCache.push(result.requestId);
});

page._client.on('Network.responseReceived', result => {
  const resp = result.response;
  resp.fromMemoryCache = responsesFromMemCache.includes(result.requestId);
  console.log(resp.url, resp.fromMemoryCache, resp.fromDiskCache, resp.fromServiceWorker);
});
```

Screenshot is from testing on https://stackoverflow.com/questions/13140318/check-whether-network-response-is-coming-from-server-or-chrome-cache:

<img width="570" alt="screen shot 2018-01-05 at 11 28 01 am" src="https://user-images.githubusercontent.com/238208/34624865-9772eb1e-f20b-11e7-8487-ff23ed635fff.png">

Want to add `response.fromMemoryCache()` or would you prefer `response.fromCache()`, which covers both cases?

aslushnikov commented:
> Does DT prioritize showing "from memory cache" if the resource is in both?

It does. However, since response could be served either from memory or disk cache, a single `response.fromCache()` makes sense to me. Later on, if there's a need, we can start returning a string instead of boolean that clarifies cache source.

ebidel committed:
fromDiskCache -> fromCache

ebidel commented:
Cool. I've changed the name.

`response.fromCache()` only considers disk cache. LMK if you want me to add the plumbing in https://github.com/GoogleChrome/puppeteer/pull/1601#issuecomment-355646010 to also consider the mem cache.

Also removed my [`Cache-Control` addition](https://github.com/GoogleChrome/puppeteer/pull/1601/commits/c701fe1eebe6cfab3dfd520f562320c59f49bfc9#diff-38e89beafd599cb655dc914a04f642feL201) to SimpleServer.js. So Travis will continue to fail until [you add caching to it](https://github.com/GoogleChrome/puppeteer/pull/1601#issuecomment-355409613).

aslushnikov commented:
> response.fromCache() only considers disk cache. LMK if you want me to add the plumbing in #1601 (comment) to also consider the mem cache.

Yeah, let's add it - otherwise it's quite hard to test the `fromCache` feature: it's hard to predict when chrome uses memcache and when it falls back to disk cache.

> Also removed my Cache-Control addition to SimpleServer.js. So Travis will continue to fail until you add caching to it.

I'll upload implementation shortly

ebidel committed:
fromCache also considers mem cache

aslushnikov commented on lib/NetworkManager.js:
```diff
@@ -302,6 +312,12 @@ class Request {
+    this._fromMemCache = false;
+  }
+
+  _setFromMemoryCache(fromMemCache) {
```
nit: let's drop the method and reach into request right away

ebidel replied:
done

ebidel commented:
LMK if that looks ok.

aslushnikov commented on lib/NetworkManager.js:
```diff
@@ -165,6 +166,15 @@ class NetworkManager extends EventEmitter {
+  _onRequestServedFromCache(event) {
+    const request = this._requestIdToRequest.get(event.requestId);
+    if (request)
+      request._setFromMemoryCache(true);
```
```js
request._fromMemoryCache = true;
```

ebidel replied:
done

aslushnikov commented on lib/NetworkManager.js:
```diff
@@ -302,6 +312,12 @@ class Request {
     for (const key of Object.keys(payload.headers))
       this._headers[key.toLowerCase()] = payload.headers[key];
+
+    this._fromMemCache = false;
```
nit: let's rename into `this._fromMemoryCache`

ebidel replied:
done

ebidel committed:
feedback

ebidel commented:
@aslushnikov ready for your additions :p

aslushnikov commented:
I wasn't able to push in this branch, so closing in favor of #1971.
